prometheus_hardware_exporter/collector.py

```diff
@@ -995,42 +995,51 @@ def fetch(self) -> List[Payload]:
-                (
-                    storage_controller_count,
-                    storage_controller_data,
-                ) = redfish_helper.get_storage_controller_data()
-                network_adapter_count: Dict[str, Any] = redfish_helper.get_network_adapter_data()
-                chassis_data: Dict[str, Dict] = redfish_helper.get_chassis_data()
-                storage_drive_count, storage_drive_data = redfish_helper.get_storage_drive_data()
-                memory_dimm_count, memory_dimm_data = redfish_helper.get_memory_dimm_data()
-                smart_storage_health_data: Dict[str, Any] = (
-                    redfish_helper.get_smart_storage_health_data()
-                )
-
+            redfish_helper = RedfishHelper(self.config)
+            redfish_helper.login()
+            payloads.append(Payload(name="redfish_call_success", value=1.0))
+
+            processor_count: Dict[str, int]
+            processor_data: Dict[str, List]
+            storage_controller_count: Dict[str, int]
+            storage_controller_data: Dict[str, List]
+            storage_drive_count: Dict[str, int]
+            storage_drive_data: Dict[str, List]
+            memory_dimm_count: Dict[str, int]
+            memory_dimm_data: Dict[str, List]
+
+            sensor_data: Dict[str, List] = redfish_helper.get_sensor_data()
+            processor_count, processor_data = redfish_helper.get_processor_data()
+            (
+                storage_controller_count,
+                storage_controller_data,
+            ) = redfish_helper.get_storage_controller_data()
+            network_adapter_count: Dict[str, Any] = redfish_helper.get_network_adapter_data()
+            chassis_data: Dict[str, Dict] = redfish_helper.get_chassis_data()
+            storage_drive_count, storage_drive_data = redfish_helper.get_storage_drive_data()
+            memory_dimm_count, memory_dimm_data = redfish_helper.get_memory_dimm_data()
+            smart_storage_health_data: Dict[str, Any] = (
+                redfish_helper.get_smart_storage_health_data()
+            )
         except (
             ConnectionError,
             InvalidCredentialsError,
             RetriesExhaustedError,
             SessionCreationError,
         ) as err:
-            logger.exception("Exception occurred while getting redfish object: %s", err)
+            logger.exception("Exception occurred while using redfish object: %s", err)
+            payloads.append(Payload(name="redfish_call_success", value=0.0))
+            return payloads
+        except Exception as err:  # pylint: disable=W0718
+            logger.exception("Exception occurred while using redfish object: %s", err)
             payloads.append(Payload(name="redfish_call_success", value=0.0))
             return payloads
+        finally:
+            # Guarding against `RedfishHelper` crashes, and `redfish_helper` remains `None`
+            if redfish_helper:
+                redfish_helper.logout()
```

Comment:
If the context manager was not logout, I believe that we should fix it there and not here. Why defining as context manager, which is not working correctly.

Reply:
The `RedfishHelper` is our object, I hope so that I'm not wrong.

Reply:
Can you check the PR description again, and see if you can understand the explanation I provided. Short answer is if `__enter__`  crashes, `__exit__` will not run. Hence, we are fixing it here

Reply:
Example:
```python3
class Foo:

    def __enter__(self):
        print("enter")
        raise KeyError

    def __exit__(self, *args, **kwargs):
        print("exit")

try:
    with Foo() as f:
        print("do something")
except ValueError:
    print("catch ValueErrror")
except Exception:
    print("catch Exception")
```

Reply:
Now I see, but still is strange behavior. Why should move that try except to the __enter__ and not outside. To me, we are not defining context manager properly.

Why we are defining redfish_obj in __enter__ and not in __init__? Like this if init failed you do not need to run logout (and try except make sense) and if login failed you should not do any logout again or?
```python
 def __init__(self, config: Config) -> None:
        """Initialize values for class."""
        self.host = config.redfish_host
        self.username = config.redfish_username
        self.password = config.redfish_password
        self.timeout = config.redfish_client_timeout
        self.max_retry = config.redfish_client_max_retry
        self.redfish_obj = redfish_client(
            base_url=self.host,
            username=self.username,
            password=self.password,
            timeout=self.timeout,
            max_retry=self.max_retry,
        )

    def __enter__(self) -> Self:
        """Login to redfish while entering context manager."""
        self.redfish_obj.login(auth="session")
        return self
```

Reply:
Yes, that's one of the option I discussed with @dashmage : adding `try-except` in the `__enter__`, but that essentially just making a nested `try-except` block. So I guess both this implementation and the one you suggested has pros and cons (in terms of the coding style, no right or wrong )